fdbrpc/include/fdbrpc/AsyncFileWriteChecker.h

```diff
@@ -35,15 +35,25 @@ class AsyncFileWriteChecker : public IAsyncFile, public ReferenceCounted<AsyncFi
 		// Lambda must hold a reference to this to keep it alive until after the read
 		auto self = Reference<AsyncFileWriteChecker>::addRef(this);
 		return map(m_f->read(data, length, offset), [self, data, offset](int r) {
-			self->updateChecksumHistory(false, offset, r, (uint8_t*)data);
+			// Do not check the checksum if self is the sole owner of the reference because the end user has dropped the
+			// file handle and may already have re-opened the file and written to it before this read completed so our
+			// stored checksum can be wrong.
+			if (!self->isSoleOwner()) {
```

Comment:
Not necessarily by this PR, but wouldn't it help to have the file handles in the simulated file system were singletons keyed by path strings? Then it'd simply suffice to have the opening of the file `FlowLock` (or something like a `FlowRWLock`) protected, which the OS does anyway to protect against concurrent access.

Reply:
It's true that the best solution would be to prevent the same file from being opened multiple times for write and make subsequent opens wait for the prior instance to be dropped, but I'm not taking on that task right now.  I think the open file map needed already exists, it's just not being used in this way and doing so is likely to hit some tricky correctness issues around process killing and restarts.

When you open a file, it's actually opened many times at once by a stack of IAsyncFiles which proxy each other.  The stack is something like:
 - `AsyncFileCached`, optional based on cached/uncached `open()` option.   This class has a static map of open files so multiple opens return the same handle.
- `AsyncFileWriteChecker`, optionally injected
- `AsyncFileChaos`, optionally injected
- `AsyncFileDetachable`
- `AsyncFileNonDurable` - These handles are stored in `g_simulator->getCurrentProcess()->machine->openFiles` by name and are not closed until the virtual process/machine is killed/rebooted.
- `SimpleFile`

The serialization point should probably be where the `AsyncFileNonDurable` handle is read from the openFiles map for the virtual process.  It  could be possible for multiple instances of any of the classes above this point in the stack to _exist_ at the same time for the same underlying file, they just would not be _used_ at the same time since the `open()` calls will wait for underlying `AsyncFileNonDurable` handle to become free.

Reply:
It turns out that this fix does not work for all cases because there can be multiple read actors open even though the user has dropped the file reference so the reference count is still > 1.

The right fix is to move AsyncFileWriteChecker lower in the stack, above SimpleFile, which is where it is created in the non-sim stack.